server/auth/auth.go

```diff
@@ -956,6 +1148,119 @@ func (auth *AuthManager) conn(conn comms.Link) *clientInfo {
+	log.Debugf("Sending tierchanged notification to %v, new tier = %d, reason = %v",
+		acctID, newTier, reason)
+	tierChangedNtfn := &msgjson.TierChangedNotification{
+		Tier:   newTier,
+		Reason: reason,
+	}
+	auth.Sign(tierChangedNtfn)
+	resp, err := msgjson.NewNotification(msgjson.TierChangeRoute, tierChangedNtfn)
+	if err != nil {
+		log.Error("TierChangeRoute encoding error: %v", err)
+		return
+	}
+	if err = auth.Send(acctID, resp); err != nil {
+		log.Warnf("Error sending tier changed notification to account %v: %v", acctID, err)
+		// The user will need to 'connect' to see their current tier and bonds.
+	}
+}
+
+// sendBondExpired sends a bondexpired notification to an account.
+func (auth *AuthManager) sendBondExpired(acctID account.AccountID, bond *db.Bond, newTier int64) {
+	log.Debugf("Sending bondexpired notification to %v for bond %v (%s), new tier = %d",
+		acctID, coinIDString(bond.AssetID, bond.CoinID), dex.BipIDSymbol(bond.AssetID), newTier)
+	bondExpNtfn := &msgjson.BondExpiredNotification{
+		AssetID:    bond.AssetID,
+		BondCoinID: bond.CoinID,
+		AccountID:  acctID[:],
+		Tier:       newTier,
+	}
+	auth.Sign(bondExpNtfn)
+	resp, err := msgjson.NewNotification(msgjson.BondExpiredRoute, bondExpNtfn)
+	if err != nil {
+		log.Error("BondExpiredRoute encoding error: %v", err)
+		return
+	}
+	if err = auth.Send(acctID, resp); err != nil {
+		log.Warnf("Error sending bond expired notification to account %v: %v", acctID, err)
+		// The user will need to 'connect' to see their current tier and bonds.
+	}
+}
+
+// checkBonds checks all connected users' bonds expiry and recomputes user tier
+// on change. This should be run on a ticker.
+func (auth *AuthManager) checkBonds() {
+	lockTimeThresh := time.Now().Add(auth.bondExpiry).Unix()
+
+	checkClientBonds := func(client *clientInfo) ([]*db.Bond, int64, int64) {
+		client.mtx.Lock()
+		defer client.mtx.Unlock()
+		pruned, bondTier := client.pruneBonds(lockTimeThresh, auth.bondAssets)
+		if len(pruned) == 0 {
+			return nil, bondTier, client.tier // no tier change
+		}
+
+		auth.violationMtx.Lock()
+		score := auth.userScore(client.acct.ID)
+		auth.violationMtx.Unlock()
+
+		client.tier = auth.tier(bondTier, score, client.legacyFeePaid)
+
+		return pruned, bondTier, client.tier
+	}
+
+	auth.connMtx.RLock()
+	defer auth.connMtx.RUnlock()
+
+	type checkRes struct {
+		tier  int64
+		bonds []*db.Bond
+	}
+	expiredBonds := make(map[account.AccountID]checkRes)
+	for acct, client := range auth.users {
+		pruned, bondTier, newTier := checkClientBonds(client)
+		if len(pruned) > 0 {
+			log.Infof("Pruned %d expired bonds for user %v, new bond tier = %d, new trading tier = %d",
+				len(pruned), acct, bondTier, client.tier)
+			expiredBonds[acct] = checkRes{newTier, pruned}
+		}
+	}
+
+	go func() { // godspeed
+		for acct, prunes := range expiredBonds {
+			for _, bond := range prunes.bonds {
+				auth.sendBondExpired(acct, bond, prunes.tier)
```

Comment:
Here's probably a good place to call a new `Archiver` method to DELETE rows in the bonds table.